selfdrive/ui/qt/onroad/model.cc

```diff
@@ -0,0 +1,208 @@
+#include "selfdrive/ui/qt/onroad/model.h"
+
+constexpr int CLIP_MARGIN = 500;
+constexpr float MIN_DRAW_DISTANCE = 10.0;
+constexpr float MAX_DRAW_DISTANCE = 100.0;
+
+static int get_path_length_idx(const cereal::XYZTData::Reader &line, const float path_height) {
+  const auto &line_x = line.getX();
+  int max_idx = 0;
+  for (int i = 1; i < line_x.size() && line_x[i] <= path_height; ++i) {
+    max_idx = i;
+  }
+  return max_idx;
+}
+
+void ModelRenderer::draw(QPainter &painter, const QRect &surface_rect) {
+  auto &sm = *(uiState()->sm);
+  if (sm.updated("carParams")) {
+    longitudinal_control = sm["carParams"].getCarParams().getOpenpilotLongitudinalControl();
+  }
+
+  // Check if data is up-to-date
+  if (!(sm.alive("liveCalibration") && sm.alive("modelV2"))) {
```

Comment:
Is that logic really equivalent to the old one? https://github.com/commaai/openpilot/blob/7fa8e3d8a65682a00d8ee6d5dba2b58292921dfc/selfdrive/ui/ui.cc#L151-L154

Reply:
It's not exactly the same, but it accomplishes the same functionality, and perhaps even better.
First, `ModelRenderer::draw()` is only called when onroad, so there’s no need to check if `scene.started` is true within `draw()`.
Second, `sm.alive(xxx)` is similar to `sm.rcv_frame("xxx") > scene.started_frame` and has the added benefit of skipping the rendering of invalid model data on the screen if the messages are already inactive for some reason. (At this point, the OnroadAlerts class may have already started displaying warnings.)